File: Minimal/Scala/url_shortener.scala
import scala.util.Random
import scala.collection.mutable

class URLShortener {
  private val urls = mutable.Map[String, String]()
  
  def shorten(url: String): String = {
    val short = generateShortCode(6)
    urls(short) = url
    short
  }
  
  def getUrl(short: String): Option[String] = {
    urls.get(short)
  }
  
  private def generateShortCode(length: Int): String = {
    val characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    (1 to length).map(_ => characters(Random.nextInt(characters.length))).mkString
  }
}

// Example usage
object URLShortenerApp extends App {
  val shortener = new URLShortener()
  val short = shortener.shorten("https://www.example.com")
  println(s"Short URL: $short")
  shortener.getUrl(short) match {
    case Some(original) => println(s"Original URL: $original")
    case None => println("URL not found")
  }
}

File: Minimal/Haskell/url_shortener.hs
import qualified Data.Map as Map
import System.Random

data URLShortener = URLShortener { urls :: Map.Map String String } deriving (Show)

emptyShortener :: URLShortener
emptyShortener = URLShortener Map.empty

shorten :: URLShortener -> String -> IO (String, URLShortener)
shorten shortener url = do
    short <- generateShortCode 6
    let newUrls = Map.insert short url (urls shortener)
    return (short, URLShortener newUrls)

getUrl :: URLShortener -> String -> Maybe String
getUrl shortener short = Map.lookup short (urls shortener)

generateShortCode :: Int -> IO String
generateShortCode n = do
    gen <- newStdGen
    let characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    return $ take n $ randomRs (0, length characters - 1) gen >>= \i -> [characters !! i]

-- Example usage
main :: IO ()
main = do
    let shortener = emptyShortener
    (short, newShortener) <- shorten shortener "https://www.example.com"
    putStrLn $ "Short URL: " ++ short
    case getUrl newShortener short of
        Just original -> putStrLn $ "Original URL: " ++ original
        Nothing -> putStrLn "URL not found"
